// @metamask/eth-json-rpc-middleware/dist/methods/wallet-get-calls-status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.walletGetCallsStatus = exports.GetCallsStatusCode = void 0;
const rpc_errors_1 = require("@metamask/rpc-errors");
const superstruct_1 = require("@metamask/superstruct");
const utils_1 = require("@metamask/utils");
const validation_1 = require("../utils/validation");
const GetCallsStatusStruct = (0, superstruct_1.tuple)([utils_1.StrictHexStruct]);
var GetCallsStatusCode;
(function (GetCallsStatusCode) {
    GetCallsStatusCode[GetCallsStatusCode["PENDING"] = 100] = "PENDING";
    GetCallsStatusCode[GetCallsStatusCode["CONFIRMED"] = 200] = "CONFIRMED";
    GetCallsStatusCode[GetCallsStatusCode["FAILED_OFFCHAIN"] = 400] = "FAILED_OFFCHAIN";
    GetCallsStatusCode[GetCallsStatusCode["REVERTED"] = 500] = "REVERTED";
    GetCallsStatusCode[GetCallsStatusCode["REVERTED_PARTIAL"] = 600] = "REVERTED_PARTIAL";
})(GetCallsStatusCode = exports.GetCallsStatusCode || (exports.GetCallsStatusCode = {}));
async function walletGetCallsStatus(req, res, { getCallsStatus, }) {
    if (!getCallsStatus) {
        throw rpc_errors_1.rpcErrors.methodNotSupported();
    }
    (0, validation_1.validateParams)(req.params, GetCallsStatusStruct);
    const id = req.params[0];
    res.result = await getCallsStatus(id, req);
}
exports.walletGetCallsStatus = walletGetCallsStatus;
